File: 6.栈与队列/8.前 K 个高频元素/347.前-k-个高频元素.cpp
/*
 * @lc app=leetcode.cn id=347 lang=cpp
 *
 * [347] 前 K 个高频元素
 */
#include <bits/stdc++.h>
using namespace std;
// @lc code=start
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        auto cmp = [](const pair<int, int>& a, const pair<int, int>& b)
        {
            return a.first > b.first;
        };
        priority_queue<pair<int, int>, vector<pair<int, int>>,decltype(cmp)> heap(cmp);
        unordered_map<int, int> hash;
        vector<int> res;
        for (auto x : nums)
        {
            hash[x] ++;
        }
        for(auto [key, val] : hash)
        {
            heap.emplace(val, key);
            if (heap.size() > k)
            {
                heap.pop();
            }
        }
        while (heap.size())
        {
            res.push_back(heap.top().second);
            heap.pop();
        }
        return res;
    }
};
// @lc code=end
class Solution {
    public:
        vector<int> topKFrequent(vector<int>& nums, int k) {
            priority_queue<pair<int, int>> heap;
            unordered_map<int, int> hash;
            vector<int> res;
            for (int i = 0; i < nums.size(); i ++ )
            {
                hash[nums[i]] ++;
            }
            for (auto [key, val] : hash)
            {
                heap.emplace(val, key);
            }
            while (k --)
            {
                res.push_back(heap.top().second);
                heap.pop();
            }
            return res;
        }
    };

File: HOT100/HOT100p5/post/347.前-k-个高频元素.py
#
# @lc app=leetcode.cn id=347 lang=python3
#
# [347] 前 K 个高频元素
#
from mytools import *
# @lc code=start
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        cnt = Counter(nums)
        n = len(nums)
        res = []
        b_ = [[] for _ in range(n + 1)]
        for k_, v in cnt.items():
            b_[v].append(k_)
        for i in range(n, -1, -1):
            res.extend(b_[i])
        while len(res) > k:
            res.pop()
        return res
# @lc code=end

